cmd/podman/images/scp.go

```diff
@@ -0,0 +1,383 @@
+	"github.com/containers/podman/v3/cmd/podman/parse"
+	"github.com/containers/podman/v3/cmd/podman/registry"
+	"github.com/containers/podman/v3/libpod/define"
+	"github.com/containers/podman/v3/pkg/domain/entities"
+	"github.com/containers/podman/v3/pkg/domain/infra"
+	t "github.com/containers/podman/v3/pkg/terminal"
+	"github.com/containers/podman/v3/pkg/util"
+	scpD "github.com/dtylman/scp"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+)
+
+var (
+	validScpFormats = []string{define.OCIManifestDir, define.OCIArchive, define.V2s2ManifestDir, define.V2s2Archive}
+)
+
+var (
+	saveScpDescription = `Securely copy an image from one host to another.`
+
+	imageScpCommand = &cobra.Command{
+		Use:   "scp [options] IMAGE [IMAGE...]",
+		Long:  saveScpDescription,
+		Short: "securely copy images",
+		RunE:  scp,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) < 1 {
+				return errors.Errorf("need at least 1 argument")
+			}
+			format, err := cmd.Flags().GetString("format")
+			if err != nil {
+				return err
+			}
+			if !util.StringInSlice(format, validScpFormats) {
+				return errors.Errorf("format value must be one of %s", strings.Join(validScpFormats, " "))
+			}
+			return nil
+		},
+		ValidArgsFunction: common.AutocompleteImages,
+		Example:           `podman image scp myimage:latest otherhost::`,
+	}
+)
+
+var (
+	scpOpts entities.ImageScpOptions
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Command: imageScpCommand,
+		Parent:  imageCmd,
+	})
+	scpFlags(imageScpCommand)
+}
+
+func scpFlags(cmd *cobra.Command) {
+	flags := cmd.Flags()
+	formatFlagName := "format"
+	flags.StringVar(&scpOpts.Save.Format, formatFlagName, define.V2s2Archive, "Save image to oci-archive, docker-archive, docker-dir")
+	_ = cmd.RegisterFlagCompletionFunc(formatFlagName, common.AutocompleteImageSaveFormat)
+	flags.BoolVarP(&scpOpts.Save.Quiet, "quiet", "q", false, "Suppress the output")
+}
+
+func scp(cmd *cobra.Command, args []string) (finalErr error) {
+	var (
+		tags []string
+		err  error
+	)
+	if scpOpts.Save.Quiet { // set quiet for both load and save
+		scpOpts.Load.Quiet = true
+	}
+	scpOpts.ToFrom = true                 // default to loading to the ssh client
+	f, err := ioutil.TempFile("", "temp") // open temp dir for load/save output
+	if err != nil {
+		return err
+	}
+	scpOpts.Save.Output = f.Name()
+	scpOpts.Load.Input = scpOpts.Save.Output
+	if err := parse.ValidateFileName(saveOpts.Output); err != nil {
+		return err
+	}
+	confR, err := config.NewConfig("") // create a hand made config for the remote engine since we might use remote and native at once
+	if err != nil {
+		return errors.Wrapf(err, "could not make config")
+	}
+	abiEng, err := registry.NewImageEngine(cmd, args) // abi native engine
+	if err != nil {
+		return errors.Wrapf(err, "could not make abi engine")
```

Comment:
Would the end user know what an abi engine is and how to check on it?

Reply:
Yes just return the err.